Modules/MUON/Common/src/TrackPlotter.cxx

```diff
@@ -0,0 +1,339 @@
+}
+
+ROOT::Math::PxPyPzMVector getMomentum4D(const o2::mch::TrackMCH& track)
+{
+  o2::mch::TrackParam trackParamAtVertex(track.getZ(), track.getParameters());
+  double vz = 0.0;
+  if (!o2::mch::TrackExtrap::extrapToVertex(trackParamAtVertex, 0.0, 0.0, 0.0, 0.0, 0.0)) {
+    std::cerr << "Track extrap failed\n";
+    return { 0, 0, 0, 0 };
+  }
+  return getMomentum4D(trackParamAtVertex);
+}
+
+namespace
+{
+struct PadSize {
+  double dx, dy;
+};
+const bool operator<(const PadSize& s1, const PadSize& s2)
+{
+  if (s1.dx == s2.dx) {
+    return s1.dy < s2.dy;
+  }
+  return s1.dx < s2.dx;
+};
+} // namespace
+
+void TrackPlotter::fillClusterHistos(gsl::span<const o2::mch::Cluster> clusters,
+                                     gsl::span<const o2::mch::Digit> digits)
+{
+
+  auto padPosCompare = [](double a, double b) {
+    constexpr double smallestPadSize = 0.21;
+    if (std::fabs(a - b) < smallestPadSize) {
+      return false;
+    }
+    return a < b;
+  };
+
+  for (const auto& cluster : clusters) {
+    int deId = cluster.getDEId();
+    const auto& seg = o2::mch::mapping::segmentation(deId);
+    int b, nb;
+    o2::math_utils::Point3D<double> global{ cluster.getX(),
+                                            cluster.getY(), cluster.getZ() };
+    auto t = mTransformation(deId).Inverse();
+    auto local = t(global);
+
+    seg.findPadPairByPosition(local.X(), local.Y(), b, nb);
+    if (b >= 0) {
+      int dsId = seg.padDualSampaId(b);
+      mNofClustersPerDualSampa->Fill(dsbinx(deId, dsId));
+    }
+    if (nb >= 0) {
+      int dsId = seg.padDualSampaId(nb);
+      mNofClustersPerDualSampa->Fill(dsbinx(deId, dsId));
+    }
+    int chamberId = cluster.getChamberId();
+    mClusterSizePerChamber->Fill(chamberId + 1, cluster.nDigits);
+    std::set<const PadSize> padSizes;
+    std::set<double, decltype(padPosCompare)> xpos(padPosCompare);
+    std::set<double, decltype(padPosCompare)> ypos(padPosCompare);
+    for (int dix = cluster.firstDigit; dix < cluster.firstDigit + cluster.nDigits; ++dix) {
+      const auto& digit = digits[dix];
+      auto padid = digit.getPadID();
+      PadSize padSize{ seg.padSizeX(padid),
+                       seg.padSizeY(padid) };
+      padSizes.emplace(padSize);
+      if (seg.isBendingPad(padid)) {
+        ypos.emplace(seg.padPositionY(padid));
+      } else {
+        xpos.emplace(seg.padPositionX(padid));
+      }
+    }
+    mClusterBendingSizePerChamber->Fill(chamberId + 1, ypos.size());
+    mClusterNonBendingSizePerChamber->Fill(chamberId + 1, xpos.size());
+  }
+}
+
+void TrackPlotter::fillTrackPairHistos(gsl::span<const o2::mch::TrackMCH> tracks)
+{
+  if (tracks.size() > 1) {
+    for (auto i = 0; i < tracks.size(); i++) {
+      auto ti = getMomentum4D(tracks[i]);
+      for (auto j = i + 1; j < tracks.size(); j++) {
+        if (tracks[i].getSign() == tracks[j].getSign()) {
+          continue;
+        }
+        auto tj = getMomentum4D(tracks[j]);
+        auto p = ti + tj;
```

Comment:
Would it make sense to add a cut on the track time difference in order to reduce the combinatorial background?

Reply:
the issue with the combinatorial background was elsewhere : I was making track pairs across the full TF, not within each ROF. Now fixed.